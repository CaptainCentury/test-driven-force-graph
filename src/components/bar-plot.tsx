import React, { FunctionComponent, useEffect, useRef } from "react";
import { max, scaleLinear, select } from "d3";

type BarPlotProps = {
  data: { label?: string; count: number }[];
  children: JSX.Element | JSX.Element[] | string;
};

export const BarPlot: FunctionComponent<BarPlotProps> = ({
  data,
  children,
  ...props
}) => {
  const svgRef = useRef(null);
  const margin = { top: 30, right: 30, bottom: 30, left: 30 };

  // define display
  var w = 300;
  var h = 300;

  useEffect(() => {
    const n = data.length;
    const maxCount = max(data, (d) => d.count);

    const svg = select("#bar-plot");
    const scaleX = scaleLinear()
      .domain([0, n - 1])
      .range([margin.left, w - margin.right]);
    const scaleY = scaleLinear()
      .domain([0, maxCount])
      .range([h - margin.bottom, margin.top]);

    const cs = svg.selectAll("line").data(data);
    cs.exit().remove();

    cs.enter().append("line").attr("stroke", "red").attr("stroke-width", 20);

    svg.exit().remove();

    svg
      .selectAll("line")
      .data(data)
      .transition()
      .duration(1000)
      .delay((d, i) => 200 * i)
      .attr("x1", (_, i) => scaleX(i + 0.5))
      .attr("x2", (_, i) => scaleX(i + 0.5))
      .transition()
      .duration(1000)
      .delay((d, i) => 200 * i)
      .attr("y1", () => scaleY(0))
      .attr("y2", (d) => scaleY(d.count));
  }, [data]);

  return (
    <figure {...props}>
      <svg
        id="bar-plot"
        ref={svgRef}
        style={{
          borderStyle: "solid",
          borderWidth: "1px",
          overflow: "visible",
        }}
        width={w + margin.left + margin.right}
        height={h + margin.top + margin.bottom}
      />
      <figcaption>
        BAR PLOT ({data ? data.length : 0} counts) {children}
      </figcaption>
    </figure>
  );
};
