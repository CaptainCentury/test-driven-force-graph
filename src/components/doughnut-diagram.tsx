import React, { FunctionComponent, useRef } from "react";

type DoughnutDiagramProps = {
  data: { name: string; votes: number }[];
  children?: string | JSX.Element | JSX.Element[];
};

export const DoughnutDiagram: FunctionComponent<DoughnutDiagramProps> = ({
  data,
  ...props
}) => {
  const svgRef = useRef(null);
  const margin = { top: 30, right: 30, bottom: 30, left: 30 };

  // define display
  var w = 300;
  var h = 300;

  return (
    <figure {...props}>
      <svg
        ref={svgRef}
        style={{ borderStyle: "solid", borderWidth: "1px" }}
        width={w + margin.left + margin.right}
        height={h + margin.top + margin.bottom}
      />
      <figcaption>DONUGHT DIAGRAM ({data.length} candidates)</figcaption>
    </figure>
  );
};
