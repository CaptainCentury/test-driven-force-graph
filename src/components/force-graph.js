import React from "react";
import { forceSimulation } from "d3-force";
import {
  drag,
  forceCenter,
  forceLink,
  forceManyBody,
  scaleOrdinal,
  schemeCategory10,
  select,
} from "d3";

export function ForceGraph({ dataset, children }) {
  const svgRef = React.useRef(null);
  const margin = { top: 30, right: 30, bottom: 30, left: 30 };

  // define display
  var w = 300;
  var h = 300;

  React.useEffect(() => {
    // init simple force layout
    var force = forceSimulation(dataset.nodes)
      .force("charge", forceManyBody())
      .force("link", forceLink(dataset.edges))
      .force(
        "center",
        forceCenter()
          .x(w / 2)
          .y(h / 2)
      );

    // display
    var svgElement = select(svgRef.current);
    svgElement.selectAll("*").remove();
    const svg = svgElement
      .append("g")
      .attr("transform", `translate(${margin.left},${margin.top})`);

    var edges = svg
      .selectAll("line")
      .data(dataset.edges)
      .enter()
      .append("line")
      .style("stroke", "#ccc")
      .style("stroke-width", 1);

    var colors = scaleOrdinal(schemeCategory10);

    var nodes = svg
      .selectAll("circle")
      .data(dataset.nodes)
      .enter()
      .append("circle")
      .attr("r", 10)
      .style("fill", function (d, i) {
        return colors(i);
      })
      .call(
        drag()
          .on("start", dragStarted)
          .on("drag", dragging)
          .on("end", dragEnded)
      );

    // simple tooltip
    nodes.append("title").text(function (d) {
      return d.name;
    });

    // simulation tick definition
    force.on("tick", function () {
      edges
        .attr("x1", function (d) {
          return d.source.x;
        })
        .attr("y1", function (d) {
          return d.source.y;
        })
        .attr("x2", function (d) {
          return d.target.x;
        })
        .attr("y2", function (d) {
          return d.target.y;
        });

      nodes
        .attr("cx", function (d) {
          return d.x;
        })
        .attr("cy", function (d) {
          return d.y;
        });
    });

    // define drag events
    function dragStarted(event, d) {
      if (!event.active) {
        force.alphaTarget(0.3).restart();
      }
      d.fx = d.x;
      d.fy = d.y;
    }

    function dragging(event, d) {
      d.fx = event.x;
      d.fy = event.y;
    }

    function dragEnded(event, d) {
      if (!event.active) {
        force.alphaTarget(0);
      }
      d.fx = null;
      d.fy = null;
    }
  }, []);

  return (
    <>
      <svg
        ref={svgRef}
        width={w + margin.left + margin.right}
        height={h + margin.top + margin.bottom}
      />
      <div>{children}</div>
    </>
  );
}
