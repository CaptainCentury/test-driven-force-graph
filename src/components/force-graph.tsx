import React, { FunctionComponent, useEffect, useRef } from "react";
import { forceSimulation } from "d3-force";
import {
  drag,
  forceCenter,
  forceLink,
  forceManyBody,
  scaleOrdinal,
  schemeCategory10,
  select,
} from "d3";

export class Visualizer {
  radius: number;
  linkStrength: number;

  constructor(radius, linkStrength) {
    this.radius = radius;
    this.linkStrength = linkStrength;
  }

  instance() {
    return new Visualizer(this.radius, this.linkStrength);
  }

  toString() {
    return `(r: ${this.radius}, link: ${this.linkStrength})`;
  }
}

type Node = {
  x: number;
  y: number;
};

type Edge = {
  name: string;
  source: Node;
  target: Node;
  x: number;
  y: number;
};

type ForceGraphProps = {
  dataset: any;
  labelMode?: string;
  visualizer?: Visualizer;
  style?: React.CSSProperties;
  children?: JSX.Element | string;
};

var colors = scaleOrdinal(schemeCategory10);

const NodeButton = (selection, radius, dragStarted, dragging, dragEnded) => {
  selection
    .append("circle")
    .attr("r", radius)
    .style("fill", function (d, i) {
      return colors(i.toString());
    });
};

const ForceGraph: FunctionComponent<ForceGraphProps> = ({
  dataset,
  labelMode = "labels",
  visualizer = new Visualizer(20, 0.01),
  children,
  ...props
}) => {
  const svgRef = useRef(null);
  const margin = { top: 30, right: 30, bottom: 30, left: 30 };

  // define display
  var w = 300;
  var h = 300;

  useEffect(() => {
    // define drag events
    function dragStarted(event, d) {
      if (!event.active) {
        force.alphaTarget(0.3).restart();
      }
      d.fx = d.x;
      d.fy = d.y;
    }

    function dragging(event, d) {
      d.fx = event.x;
      d.fy = event.y;
    }

    function dragEnded(event, d) {
      if (!event.active) {
        force.alphaTarget(0);
      }
      d.fx = null;
      d.fy = null;
    }

    // init simple force layout
    var force = forceSimulation(dataset.nodes)
      .force("charge", forceManyBody())
      .force("link", forceLink(dataset.edges).strength(visualizer.linkStrength))
      .force(
        "center",
        forceCenter()
          .x(w / 2)
          .y(h / 2)
      );

    // display
    var svgElement = select(svgRef.current);
    svgElement.selectAll("*").remove();
    const svg = svgElement
      .append("g")
      .attr("transform", `translate(${margin.left},${margin.top})`);

    var edges = svg
      .selectAll("line")
      .data(dataset.edges)
      .enter()
      .append("line")
      .style("stroke", "#ccc")
      .style("stroke-width", 1);

    const radius = visualizer.radius;

    var nodes = svg
      .selectAll("g")
      .data(dataset.nodes)
      .enter()
      .append("g")
      .call(NodeButton, radius, dragStarted, dragging, dragEnded)
      .call(
        drag()
          .on("start", dragStarted)
          .on("drag", dragging)
          .on("end", dragEnded)
      );

    const fontSize = 8;

    var nodeLabels =
      labelMode == "labels" &&
      svg
        .selectAll(".nodelabel")
        .data<Edge>(dataset.nodes)
        .enter()
        .append("text")
        .attr("text-anchor", "middle")
        .attr("dy", "0.3em")
        .attr("x", function (d) {
          return d.x;
        })
        .attr("y", function (d) {
          return d.y;
        })
        .attr("class", "nodelabel")
        .attr("fill", "#fff")
        .attr("stoke", "none")
        .text(function (d) {
          return d.name;
        })
        .style("font-size", function (d) {
          return (
            Math.min(
              2 * radius,
              ((2 * radius - 1) / this.getComputedTextLength()) * fontSize
            ) + "px"
          );
        })
        .attr("pointer-events", "none");

    // simple tooltip

    if (labelMode == "tooltip") {
      nodes.append("title").text(function (d: { name: string }) {
        return d.name;
      });
    }

    // simulation tick definition
    force.on("tick", function () {
      nodes.attr("transform", function (d: Node) {
        return `translate(${d.x}, ${d.y})`;
      });

      edges
        .attr("x1", function (d: Edge) {
          return d.source.x;
        })
        .attr("y1", function (d: Edge) {
          return d.source.y;
        })
        .attr("x2", function (d: Edge) {
          return d.target.x;
        })
        .attr("y2", function (d: Edge) {
          return d.target.y;
        });

      if (nodeLabels) {
        nodeLabels
          .attr("x", function (d) {
            return d.x;
          })
          .attr("y", function (d) {
            return d.y;
          });
      }
    });
  }, [dataset, visualizer]);

  return (
    <figure {...props}>
      <svg
        ref={svgRef}
        style={{ borderStyle: "solid", borderWidth: "1px" }}
        width={w + margin.left + margin.right}
        height={h + margin.top + margin.bottom}
      />
      <figcaption>{children}</figcaption>
    </figure>
  );
};

export default ForceGraph;
