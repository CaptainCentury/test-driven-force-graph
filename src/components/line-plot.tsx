import React, { FunctionComponent, useEffect, useRef } from "react";
import d3, { extent, line, scaleLinear, select } from "d3";

type LinePlotProps = {
  dataTable: d3.DSVRowArray<string>;
};

export const LinePlot: FunctionComponent<LinePlotProps> = ({
  dataTable,
  ...props
}) => {
  const svgRef = useRef(null);
  const margin = { top: 30, right: 30, bottom: 30, left: 30 };

  // define display
  var w = 300;
  var h = 300;

  const colors = ["green", "red"];

  useEffect(() => {
    if (dataTable && dataTable.length > 0) {
      const labels = Object.keys(dataTable[0]);

      var scaleX = scaleLinear()
        .domain(extent(dataTable, (d) => Number(d["x"])))
        .range([0, w]);

      for (let i = 1; i < labels.length; i++) {
        var scaleY = scaleLinear()
          .domain(extent(dataTable, (d) => Number(d[labels[i]])))
          .range([h, 0]);

        select("#line-plot")
          .append("g")
          .attr("id", `line${i}`)
          .selectAll("circle")
          .data(dataTable)
          .enter()
          .append("circle")
          .attr("r", 5)
          .attr("fill", colors[i - 1])
          .attr("cx", (d) => scaleX(Number(d[labels[0]])))
          .attr("cy", (d) => scaleY(Number(d[labels[i]])));

        const lineMaker = line()
          .x((d) => scaleX(d[labels[0]]))
          .y((d) => scaleY(d[labels[i]]));

        select(`#line${i}`)
          .append("path")
          .attr("fill", "none")
          .attr("stroke", colors[i - 1])
          .attr("d", lineMaker(dataTable as Iterable<[number, number]>));
      }
    }
  }, [dataTable]);

  return (
    <figure {...props}>
      <svg
        id="line-plot"
        ref={svgRef}
        style={{ borderStyle: "solid", borderWidth: "1px" }}
        width={w + margin.left + margin.right}
        height={h + margin.top + margin.bottom}
      />
      <figcaption>
        LINE PLOT ({dataTable ? dataTable.length : 0} rows)
      </figcaption>
    </figure>
  );
};
