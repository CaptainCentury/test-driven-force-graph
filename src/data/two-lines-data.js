import d3 from "d3";

export const twoLinesData = d3.tsv("./two-lines-data.tsv");
