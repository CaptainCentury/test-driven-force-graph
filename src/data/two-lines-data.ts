import { tsv } from "d3";

export const twoLinesDataPromise = tsv(`two-lines-table.tsv`);
