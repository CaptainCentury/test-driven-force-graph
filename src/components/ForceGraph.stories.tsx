import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";

import ForceGraph, { Visualizer } from "./force-graph";
import namesData from "../data/names";
import abcData from "../data/abc";
import { yarnball } from "../data/network-helper";

export default {
  title: "Example/ForceGraph",
  component: ForceGraph,
  argTypes: {
    labelMode: { control: "select", options: ["labels", "tooltip"] },
  },
} as ComponentMeta<typeof ForceGraph>;

const Template: ComponentStory<typeof ForceGraph> = (args) => (
  <ForceGraph {...args} />
);

export const abc = Template.bind({});
abc.args = {
  style: { color: "blue" },
  visualizer: new Visualizer(30, 1),
  dataset: abcData,
  labelMode: "labels",
};

export const names = Template.bind({});
names.args = {
  style: { color: "blue" },
  visualizer: new Visualizer(30, 1),
  dataset: namesData,
  labelMode: "labels",
};

export const yarn = Template.bind({});
yarn.args = {
  style: { color: "blue" },
  visualizer: new Visualizer(30, 1),
  dataset: yarnball,
  labelMode: "labels",
};
