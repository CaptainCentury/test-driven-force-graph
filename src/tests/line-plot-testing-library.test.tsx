/**
 * @jest-environment jsdom
 */
import React from "react";
import { LinePlot } from "../components/line-plot";
import { render } from "@testing-library/react";
import { twoLinesDataPromise } from "../data/two-lines-data";
import { tsvParse } from "d3";

describe("LinePlot", () => {
  let data = tsvParse("x1\ty1\ty2\n1.0\t0.5\t0.25");

  it("renders", () => {
    const testId = "plot1";
    const { getByTestId } = render(
      <LinePlot data-testid={testId} dataTable={data} />
    );
    expect(getByTestId(testId).tagName).toBe("FIGURE");
  });
});
