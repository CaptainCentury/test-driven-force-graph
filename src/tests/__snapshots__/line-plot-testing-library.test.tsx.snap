// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinePlot renders 1`] = `
<figure
  data-testid="plot1"
>
  <svg
    cursor="crosshair"
    height="360"
    id="line-plot"
    style="border-style: solid; border-width: 1px; overflow: visible;"
    width="360"
  >
    <g
      id="line1"
    >
      <circle
        cx="30"
        cy="150"
        fill="green"
        r="5"
      />
      <circle
        cx="330"
        cy="150"
        fill="green"
        r="5"
      />
      <path
        d="M30,150L180,150L180,150L330,150"
        fill="none"
        stroke="green"
      />
    </g>
    <g
      id="line2"
    >
      <circle
        cx="30"
        cy="260.7692307692307"
        fill="red"
        r="5"
      />
      <circle
        cx="330"
        cy="30"
        fill="red"
        r="5"
      />
      <path
        d="M30,260.7692307692307L180,260.7692307692307L180,30L330,30"
        fill="none"
        stroke="red"
      />
    </g>
    <g
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="middle"
      transform="translate(0, 330)"
    >
      <path
        class="domain"
        d="M30.5,6V0.5H330.5V6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(30.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          1.0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(60.50000000000003,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          1.1
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(90.49999999999999,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          1.2
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(120.50000000000001,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          1.3
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(150.49999999999997,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          1.4
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(180.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          1.5
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(210.50000000000003,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          1.6
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(240.49999999999997,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          1.7
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(270.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          1.8
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(300.49999999999994,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          1.9
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(330.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          2.0
        </text>
      </g>
    </g>
    <g
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="end"
      transform="translate(30, 0)"
    >
      <path
        class="domain"
        d="M-6,330.5H0.5V30.5H-6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(0,180.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          1
        </text>
      </g>
    </g>
    <text />
  </svg>
  <figcaption>
    LINE PLOT (
    2
     rows) 
  </figcaption>
</figure>
`;

exports[`LinePlot renders for empty data 1`] = `
<figure
  data-testid="plot1"
>
  <svg
    height="360"
    id="line-plot"
    style="border-style: solid; border-width: 1px; overflow: visible;"
    width="360"
  />
  <figcaption>
    LINE PLOT (
    0
     rows) 
  </figcaption>
</figure>
`;

exports[`LinePlot renders for undefined data 1`] = `
<figure
  data-testid="plot1"
>
  <svg
    height="360"
    id="line-plot"
    style="border-style: solid; border-width: 1px; overflow: visible;"
    width="360"
  />
  <figcaption>
    LINE PLOT (
    0
     rows) 
  </figcaption>
</figure>
`;
